syntax = "proto3";

package kv;

service KVService {
  rpc Put (PutRequest) returns (PutResponse);
  rpc Get (GetRequest) returns (GetResponse);
}

service ReplicationService {
  rpc Replicate (ReplicationPacket) returns (ReplicationAck);
}

message PutRequest {
  string key = 1;
  string value = 2;
}

message PutResponse {
  bool success = 1;
  string leader_target = 2;
}

message GetRequest {
  string key = 1;
}

message GetResponse {
  bool found = 1;
  string value = 2;
}

message Operation {
  int64 index = 1;
  string key = 2;
  string value = 3;
}

message ReplicationPacket {
  int64 from_index = 1;
  repeated Operation ops = 2;
  int64 commit_index = 3;   // NEW
}

message ReplicationAck {
  bool success = 1;
  int64 last_index = 2;
}